tephra-core/src/main/java/co/cask/tephra/persist/HDFSTransactionLog.java

```diff
@@ -157,11 +224,69 @@ public TransactionEdit next(TransactionEdit reuse) throws IOException {
       if (closed) {
         return null;
       }
-      boolean successful = reader.next(key, reuse);
-      if (successful) {
-        return reuse;
+      if (transactionEdits.size() != 0) {
+        return transactionEdits.remove(0);
+      } else {
+        if (newVersion) {
+          // v2 format, buffer edits till we reach a marker
+          boolean successful;
+          DataOutputBuffer rawKey = new DataOutputBuffer();
+
+          if (reader.nextRawKey(rawKey) != -1) {
+            // has data and has not reached EOF
+            byte [] keyBytes = TxConstants.TransactionLog.NUM_ENTRIES_APPENDED.getBytes();
+
+            if (rawKey.getLength() == keyBytes.length &&
+              Arrays.equals(Arrays.copyOf(rawKey.getData(), rawKey.getLength()), keyBytes)) {
+              int numEntries;
+              SequenceFile.ValueBytes valueBytes = reader.createValueBytes();
+              Closeables.closeQuietly(rawKey);
+
+              // key matched, read the data now to determine numEntries to read.
+              reader.nextRawValue(valueBytes);
+              ByteArrayOutputStream value = new ByteArrayOutputStream(Integer.SIZE);
+              DataOutputStream outputStream = new DataOutputStream(value);
+              valueBytes.writeUncompressedBytes(outputStream);
+              outputStream.flush();
+              Closeables.closeQuietly(outputStream);
+              numEntries = ByteBuffer.wrap(value.toByteArray()).getInt();
+              Closeables.closeQuietly(value);
+
+              // read numEntries into the list
+              for (int i = 0; i < numEntries; i++) {
+                TransactionEdit edit = new TransactionEdit();
+                try {
+                  successful = reader.next(key, edit);
+                  if (successful) {
+                    transactionEdits.add(edit);
+                  } else {
+                    transactionEdits.clear();
+                    return null;
+                  }
+                } catch (EOFException e) {
+                  // if we have reached EOF before reading back the numEntries, we clear the partial list and return.
+                  transactionEdits.clear();
+                  return null;
+                }
+              }
+            } else {
+              LOG.error("Invalid key for num entries appended found, expected : {}",
+                        TxConstants.TransactionLog.NUM_ENTRIES_APPENDED);
+            }
+          }
+
+          if (!transactionEdits.isEmpty()) {
+            return transactionEdits.remove(0);
+          } else {
+            return null;
+          }
+
+        } else {
+          // does not have marker, version v1.
+          boolean successful = reader.next(key, reuse);
```

Comment:
I think you still need to handle exception in this case so that if you encountered an old corrupted log, the last corrupted entry is getting skipped.